backend/src/benchmark/block_dev.rs

```diff
@@ -0,0 +1,50 @@
+use std::{
+    fs,
+    io::{self, Read},
+    os::unix::fs::OpenOptionsExt,
+    path::{Path, PathBuf},
+    ptr, slice, time,
+};
+
+#[derive(Eq, Ord, PartialEq, PartialOrd)]
+pub struct BlockDev(PathBuf);
+
+impl BlockDev {
+    pub fn new(path: PathBuf) -> Self {
+        Self(path)
+    }
+
+    pub fn path(&self) -> &Path {
+        &self.0
+    }
+
+    pub fn benchmark(&self) -> io::Result<f64> {
+        let mut file = fs::OpenOptions::new()
+            .read(true)
+            .custom_flags(libc::O_DIRECT)
+            .open(self.path())?;
+
+        // Buffer needs to be aligned for direct reads
+        let mut ptr = ptr::null_mut();
+        let align = 4096;
+        let size = align * 1024;
+        let res = unsafe { libc::posix_memalign(&mut ptr, align, size) };
+        if res != 0 {
+            return Err(io::Error::from_raw_os_error(res));
+        }
+
+        let elapsed = {
+            let data = unsafe { slice::from_raw_parts_mut(ptr as *mut u8, size) };
+
+            let start = time::Instant::now();
+            file.read(data)?;
```

Comment:
Presumably an error here will leak `data`.

Wonder if there's a better way to allocate here... I guess `std::alloc::alloc` would work, but have the same semantics and not really be any better.

Reply:
That's true, I'll fix this. There's not really any smart way to do page aligned allocations, but I could certainly make an abstraction that has drop implemented